### DISCUSSION TITLE
IntelliSense for Content Collection JSON

### DISCUSSION THREAD

**Discussion by joe-bell** [withastro/roadmap#696] [Proposal]
# Summary

Auto-generate Content Collection JSON [`$schema`](https://json-schema.org/understanding-json-schema/reference/schema.html#:~:text=The%20%24schema%20keyword%20is%20used,to%20the%20dialect%20%24schema%20identifies.)s for TypeScript-style autocompletion/hints (in tools like VSCode)

# Background & Motivation

I've been thoroughly enjoying using `'data'`-based Content Collections within Astro, but sorely missing the immediate auto-completion/hints that come when writing TypeScript

I recently had to edit a `vercel.json` file, which starts with the following:

```json
{ 
  "$schema": "https://openapi.vercel.sh/vercel.json"
}
```

...and found the DX within VSCode incredibly valuable. I didn't have to jump to documentation or a config file, I just let VSCode do its thing!

Then it occurred to me: could Astro do the same?

```json
{ 
  "$schema": "./.astro/$schema/collections/my-collection-name.json"
}
```

# Goals

- Enable autocompletion/hinting of JSON data within tools like VSCode
- Expose local `$schema` files to reference within JSON data

# Example

-  https://iancarpenter.dev/2020/03/01/how-to-reference-a-local-json-schema/

---

> **Note**
> I would also be open to helping with implementation

**Comment by jasikpark** [comment]
another useful ref is the `zod-to-json-schema` tools in https://zod.dev/?id=zod-to-x

**Comment by natemoo-re** [comment]
Oh wow I really had not considered this. What a wonderful idea, I'd be super interested in making this happen!

**Comment by coding-in-public** [comment]
I'd pay good money for this! How can we make this happen? I'm happy to help however I can.

**Comment by Princesseuh** [comment]
Doesn't seem too hard to generate the schemas (using the library Caleb shared), but I'm not sure what's the best way to hook the schema to the files

There's multiple ways:
- Edit the `.json` files for the user to add the schema (Works everywhere, modifying files in the user's repo is always risky)
- Edit the repo's `.vscode/settings.json` [with links to the workspace schemas](https://code.visualstudio.com/Docs/languages/json#_mapping-to-a-schema-in-the-workspace) (only works in VS Code, modifying files in the user's repo is always risky)
- Our language-server enables itself inside JSON files and do the work virtually (takes the most amount of work, doesn't work in other editors without our tooling (though it's unlikely for people to have a dedicated JSON editor))

Generating the JSON schemas would already be pretty great, so at least users can hook it themselves in the meantime while we figure out the best way to hook it automatically

**Comment by alexanderniebuhr** [comment]
I started implementing this here: https://github.com/withastro/astro/pull/10145

**Reply by natemoo-re** [reply]
Exciting stuff, thanks for kicking this off @alexanderniebuhr!

For context, I don't think this necessarily needs to go through the full RFC process since it's an incremental DX improvement on our existing APIs. But I'm willing to discuss that if anyone disagrees and would find value in a full design discussion!

**Comment by ematipico** [comment]
Closing, this is landed as experimental in 4.5

**Reply by ImLunaHey** [reply]
can you confirm if this works with data types json/yml or only markdown files?

**Reply by delucis** [reply]
I believe it works with JSON currently: https://docs.astro.build/en/guides/content-collections/#enabling-json-schema-generation
